<?xml version="1.0"?>
<launch>

	<!-- arguments -->
	<arg name="domain_path"		default="$(find rosplan_demos)/common/iros_problems_free/domain_robot_delivery.pddl" />
	<arg name="problem_path"	default="$(find rosplan_demos)/common/iros_problems_free/problem_w5_m5_g3.pddl" />

	<!-- knowledge base -->
	<node name="rosplan_knowledge_base" pkg="rosplan_knowledge_base" type="knowledgeBase" respawn="false" output="screen">
		<param name="domain_path" value="$(arg domain_path)" />
		<param name="problem_path" value="$(arg problem_path)" />
		<param name="database_path" value="$(find rosplan_knowledge_base)/common/mongoDB/" />
		<!-- conditional planning flags -->
		<param name="use_unknowns" value="false" />
	</node>

	<!-- problem generation -->
	<include file="$(find rosplan_planning_system)/launch/includes/problem_interface.launch">
		<arg name="knowledge_base"   value="rosplan_knowledge_base" />
		<arg name="domain_path"      value="$(arg domain_path)" />
		<arg name="problem_path"     value="$(find rosplan_demos)/common/problem.pddl" />
		<arg name="problem_topic"    value="problem_instance" />
	</include>

	<!-- planner interface -->
	<include file="$(find rosplan_planning_system)/launch/includes/planner_interface.launch">
		<arg name="use_problem_topic"    value="true" />
		<arg name="problem_topic"        value="/rosplan_problem_interface/problem_instance" />
		<arg name="planner_topic"        value="planner_output" />
		<arg name="domain_path"          value="$(arg domain_path)" />
		<arg name="problem_path"         value="$(find rosplan_demos)/common/problem.pddl" />
		<arg name="data_path"            value="$(find rosplan_demos)/common/" />
		<arg name="planner_command"      value="timeout 20 $(find rosplan_planning_system)/common/bin/popf DOMAIN PROBLEM" />
	</include>

	<!-- plan parsing -->
	<include file="$(find rosplan_planning_system)/launch/includes/parsing_interface.launch">
		<arg name="knowledge_base"   value="rosplan_knowledge_base" />
		<arg name="planner_topic"    value="/rosplan_planner_interface/planner_output" />
		<arg name="plan_topic"       value="complete_plan" />
	</include>

	<!-- CSP exec -->
	<node name="csp_exec_generator" pkg="rosplan_planning_system" type="csp_exec_generator" respawn="false" output="screen">
	</node>

	<!-- plan dispatching -->
	<node name="rosplan_plan_dispatcher" pkg="rosplan_planning_system" type="pddl_adaptable_plan_dispatcher" respawn="false" output="screen">
		<param name="knowledge_base"        value="rosplan_knowledge_base" />
		<param name="plan_topic"            value="/csp_exec_generator/valid_plans" />
		<param name="action_dispatch_topic" value="action_dispatch" />
		<param name="action_feedback_topic" value="action_feedback" />
		<param name="display_edge_type"     value="false" />
	</node>

	<!-- sim actions -->
    <include file="$(find rosplan_planning_system)/launch/includes/simulated_action.launch" >
        <arg name="pddl_action_name" value="goto_waypoint" />
        <arg name="action_duration" value="1" />
    </include>

    <include file="$(find rosplan_planning_system)/launch/includes/simulated_action.launch" >
        <arg name="pddl_action_name" value="localise" />
        <arg name="action_duration" value="1" />
    </include>

    <include file="$(find rosplan_planning_system)/launch/includes/simulated_action.launch" >
        <arg name="pddl_action_name" value="dock" />
        <arg name="action_duration" value="1" />
    </include>

    <include file="$(find rosplan_planning_system)/launch/includes/simulated_action.launch" >
        <arg name="pddl_action_name" value="undock" />
        <arg name="action_duration" value="1" />
    </include>

    <include file="$(find rosplan_planning_system)/launch/includes/simulated_action.launch" >
        <arg name="pddl_action_name" value="wait_load_at_machine" />
        <arg name="action_duration" value="1" />
    </include>

    <include file="$(find rosplan_planning_system)/launch/includes/simulated_action.launch" >
        <arg name="pddl_action_name" value="switch_machine_on" />
        <arg name="action_duration" value="1" />
    </include>

    <include file="$(find rosplan_planning_system)/launch/includes/simulated_action.launch" >
        <arg name="pddl_action_name" value="wait_unload" />
        <arg name="action_duration" value="1" />
    </include>

    <include file="$(find rosplan_planning_system)/launch/includes/simulated_action.launch" >
        <arg name="pddl_action_name" value="ditch" />
        <arg name="action_duration" value="1" />
    </include></launch>

